packages/compass-components/src/components/guide-cue/guide-cue.tsx

```diff
@@ -0,0 +1,222 @@
+import React, {
+  useEffect,
+  useCallback,
+  useRef,
+  useMemo,
+  useState,
+} from 'react';
+import { guideCueService, type ShowCueEventDetail } from './guide-cue-service';
+import { GuideCue as LGGuideCue } from '@leafygreen-ui/guide-cue';
+import type { GroupName } from './guide-cue-groups';
+import { css, cx } from '../..';
+import { rafraf } from '../../utils/rafraf';
+
+const hiddenPopoverStyles = css({
+  display: 'none !important',
+  opacity: '0 !important',
+  transition: 'none !important',
+});
+
+const getDataCueId = ({
+  cueId,
+  groupId,
+}: {
+  cueId: string;
+  groupId?: string;
+}): string => {
+  return `guide-cue-${cueId}${groupId ? `-${groupId}` : ''}`;
+};
+
+type LGGuideCueProps = React.ComponentProps<typeof LGGuideCue>;
+
+type GroupAndStep =
+  | {
+      groupId: GroupName;
+      step: number;
+    }
+  | {
+      groupId?: never;
+      step?: never;
+    };
+
+// omit the props we are handling
+export type GuideCueProps<T> = Omit<
+  LGGuideCueProps,
+  | 'currentStep'
+  | 'refEl'
+  | 'numberOfSteps'
+  | 'onDismiss'
+  | 'open'
+  | 'onPrimaryButtonClick'
+  | 'setOpen'
+> &
+  GroupAndStep & {
+    cueId: string;
+    trigger: ({ ref }: { ref: React.Ref<T> }) => React.ReactElement;
+  };
+
+export const GuideCue = <T extends HTMLElement>({
+  children,
+  trigger,
+  cueId,
+  groupId,
+  step,
+  ...restOfCueProps
+}: React.PropsWithChildren<GuideCueProps<T>>) => {
+  const [isOpen, setIsOpen] = useState(false);
+  const intersectionRef = useRef(true);
+  const refEl = useRef<T>(null);
+  const [readyToRender, setReadyToRender] = useState(false);
+
+  const cueData = useMemo(() => {
+    if (!groupId) {
+      return { cueId, step: 1 };
+    }
+    return { cueId, groupId, step };
+  }, [cueId, groupId, step]);
+
+  const observerCallback = useCallback(
+    (entries: IntersectionObserverEntry[]) => {
+      const [entry] = entries;
+      intersectionRef.current = entry.isIntersecting;
+
+      if (!entry.isIntersecting) {
+        setIsOpen(false);
+      }
+
+      guideCueService.onCueIntersectionChange(
+        entry.isIntersecting,
+        cueData.cueId,
+        cueData.groupId
+      );
+    },
+    [intersectionRef]
+  );
+
+  useEffect(() => {
+    if (!refEl.current) {
+      return;
+    }
+    const node = refEl.current;
+    const observer = new IntersectionObserver(observerCallback, {
+      threshold: 0.5,
+    });
+    observer.observe(node);
+    return () => {
+      observer.unobserve(node);
+      observer.disconnect();
+    };
+  }, [refEl, observerCallback]);
```

Comment:
```suggestion
  }, [observerCallback]);
```

(also if observerCallback is only used inside this effect, you can totally just move the definition there)

Reply:
good point. I moved it in there.